File: solutions/csharp/luhn/1/Luhn.cs
public static class Luhn {
  public static bool IsValid(string number) {
    if (number.Any(c => !char.IsDigit(c) && c != ' ')) {
      return false;
    }

    var cleanedNumber = number
      .Where(char.IsDigit)
      .Select(c => int.Parse(c.ToString()))
      .ToArray();

    if (cleanedNumber.Count() <= 1) {
      return false;
    }

    for (var i = cleanedNumber.Length - 2; i >= 0; i -= 2) {
      cleanedNumber[i] *= 2;
      while (cleanedNumber[i] > 9) {
        cleanedNumber[i] -= 9;
      }
    }

    return cleanedNumber.Sum() % 10 == 0;
  }
}

File: solutions/python/luhn/1/luhn.py
class Luhn:
    def __init__(self, card_num):
        self.card_num = card_num

    def valid(self):
        if any(not x.isdigit() for x in self.card_num.replace(" ", "")):
            return False
        if len(self.card_num.replace(" ", "")) <= 1:
            return False
            
        cleaned_card_num = [int(x) for x in self.card_num.replace(" ", "")]
        for num in range(len(cleaned_card_num) - 2, -1, -2):
            cleaned_card_num[num] = cleaned_card_num[num] * 2 - (9 if cleaned_card_num[num] * 2 > 9 else 0)
        return sum(cleaned_card_num) % 10 == 0
